app/javascript/utils/customHooks.js

```diff
@@ -132,3 +134,10 @@ export const useStateObject = (initialStateObject) => {
 
   return [state, setState, overrideEntireState]
 }
+
+/**
+ * Use this instead of useContext(AppContext) directly. The only reason this is better is it
+ * provides a default that will destructure nicely when useContext returns false, like when
+ * shallow rendering a component.
+ */
+export const useAppContext = () => useContext(AppContext) ?? [undefined, () => {}]
```

Comment:
One thing that's annoying is any component that uses context state can't be shallow rendered. However, if it doesn't use the state and only uses the `dispatch` function like this:
```
const [, dispatch] = useAppContext()
```
you can shallow render it (but dispatch won't do anything/be spyable unless you mount render it)